### DISCUSSION TITLE
Expose interface for foxess_em to use

### DISCUSSION THREAD

**Discussion by canton7** [nathanmarlor/foxess_modbus#373] [Ideas]
Hi @nathanmarlor,

I'm thinking about the coming winter and possibly needing to make use of foxess_em. Since I use the W610 with UDP that means I can either add UDP support to foxess_em, or...

I was wondering whether it would be easier/better to instead expose an interface from foxess_modbus which other integrations (such as foxess_em) to:

1. Query a list of configured inverters
2. Using a friendly name / device ID, fetch an instance of an interface which allows them to:
    a. Fetch the entity_id of an entity with a particular key (abstracting away the entity_id_prefix stuff)
    b. Set charge periods (or maybe they can just call the service, might be easier)
    c. Set registers? Maybe this isn't necessary if they can interact with number entities / the charge period service / etc
    d. Possibly get notified when an entity changes? You can subscribe to this using hass.

This would be made available in `hass.data`.

This means:

1. We don't have to keep foxess_em up to feature parity with foxess_modbus re connection types
2. We don't have problems with foxess_em and foxess_modbus trying to talk to the adapter at the same time
3. We don't need to double-poll the entities which foxess_em cares about

**Comment by nathanmarlor** [comment]
@canton7 Absolutely. It's been on my to-do list for a while to migrate EM to use the Modbus services directly - piggybacking on the great work you've been doing - to improve compatibility, reliability etc. The Modbus service is now lightyears ahead...! 

EM was my first foray into Python development, and whilst it's stable and functions well there are a lot of things I'd improve given the chance...

- type checking with mypi (XL)
- clean up of Solcast API calls (S)
- explore using ARIMA models for house load (M)
- revisit/refactor how the battery model is built and persisted (i.e use local storage Vs entity attributes) (M)
- basic unit tests to cover HA upgrades (M)

As always I'd appreciate any contributions you can make and would be happy to set you as a contributor as with Modbus :)

**Reply by canton7** [reply]
Cool! 

I think we might need a pypi package with some interfaces in perhaps? So foxess_em doesn't have to take a direct dependency on foxess_modbus.

**Reply by canton7** [reply]
Thanks for the steer on ARIMA - I've been trying out various things (quite naively, granted) without much success. SARIMA with (1, 1, 1) and (1, 1, 1, 24) seems to work fairly well though.

I'd be interested to see how it compares to what you're doing in EM. 

I want to do some k-means grouping across days of the week and try training different models for each group - that should help account for eg weekends having higher usage.

The aim is to feed that, and solcast data, into a model of the inverter+battery system (modelling soc, import, export). Let the model control work mode + max and min soc, and optimise it for profit. Re-run every hour if necessary. That should handle things like flux, power-ups and even agile. It's looking promising so far, but it'll likely be spring before I have something that's worth sharing widely, if at all.

If you're up for throwing some ideas around that might be useful - happy for you to take any findings to improve EM.
